File: algorithms/graph/kruskal.cpp
#include<bits/stdc++.h>
using namespace std;
#define PIII pair<int, pair<int, int>> // {weight, {x,y}}

vector<int> id;

int root(int x){
  while(id[x] != x){
    id[x] = id[id[x]];
    x = id[x];
  }
  return x;
}

void union_(int x, int y){
  int X = root(x);
  int Y = root(y);
  id[X] = id[Y];
}

int kruskal(vector<PIII> graph){
  int mincost = 0;
  for(int i=0; i<graph.size(); i++){
    int x = graph[i].second.first;
    int y = graph[i].second.second;
    int cost = graph[i].first;

    if(root(x) != root(y)){
      mincost += cost;
      union_(x, y);
    }
  }
  return mincost;
}

int main(){
  vector<PIII> graph = { {4,{0,1}}, {1,{0,3}}, {3,{1,2}}, {2,{3,2}}, {2,{2,4}} };
  sort(graph.begin(), graph.end());
  for(int i=0; i<graph.size(); i++) id.push_back(i);
  cout << kruskal(graph);
}

File: algorithms/graph/kruskal.py
import heapq

def root(id, x):
  while id[x] != x:
    id[x] = id[id[x]]
    x = id[x]

  return x

def union(id, x, y):
  print(id)
  id[root(id, x)] = id[root(id, y)]

def kruskal(graph):
  id = list(range(len(graph)))
  graph.sort(key=lambda g: g[0])
  mst = 0 # minimum spanning tree
  for d, x, y in graph:
    if root(id, x) != root(id, y):
      mst += d
      union(id, x, y)

  return mst

graph =[ [4,0,1], [1,0,3], [3,1,2], [2,3,2], [2,2,4]] # [weight, x, y]
mst = kruskal(graph)
print(mst)
